### PR TITLE
M2P-244 Collect Magento discounts with no coupon properly

### PR DESCRIPTION
**Pull request by dumega**
# Description
Currently if any Cart Price Rule with No Coupon is applied to the cart, it combines with the coupon code (which is Cart Price Rule with Specific Coupon ), so we need to separate these sale rules from each other.

image
image

Fixes: https://boltpay.atlassian.net/browse/M2P-244

#changelog Collect Magento discounts with no coupon properly

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [x] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Plugin/SalesRuleQuoteDiscountPlugin.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    
+    public function __construct(
+        CheckoutSession $checkoutSession
+    ) {
+        $this->checkoutSession = $checkoutSession;
+    }
+    
+    public function beforeCollect(\Magento\SalesRule\Model\Quote\Discount $subject, $quote, $shippingAssignment, $total)
+    {
+        $this->checkoutSession->setBoltCollectSaleRuleDiscounts([]);
```
Comment (vitaliyreznikov) [inline Plugin/SalesRuleQuoteDiscountPlugin.php:34]: @dumega When this actually happens?

If logged in user adds a coupon, logout and login again, will it work?

**Review by vitaliyreznikov — COMMENTED** [review]
Thank you for noticing this, a very important issue.
Added one question.

File: Plugin/SalesRuleQuoteDiscountPlugin.php (hunk 1 above)
Comment (dumega) [reply Plugin/SalesRuleQuoteDiscountPlugin.php:36]: @vitaliyreznikov good catch/. there is a bug when logout/login. I will look into it.

Comment (dumega) [reply Plugin/SalesRuleQuoteDiscountPlugin.php:36]: @vitaliyreznikov After investigation, this is a bug with Amasty reward point (it is enabled on my dev site), and I have created a commit to fix it.
And for your concern, Mangeto would collect the address discount amount every time when refresh the cart page or via ajax, so this session data can be reset properly.

**Review by danac-gs — APPROVED** [review]

**Review by vitaliyreznikov — APPROVED** [review]

**Pull request merged**